# Part 9 - Very Useful Pandas Functionalities
by Kaan Kabalak @ witfuldata.com

The Pandas data analysis library has many functionalities. In this part we will take a look at some of the most useful ones.

In [1]:
import numpy as np
import pandas as pd

store_df = pd.read_csv("online_retail.csv")


## Check the length of the data frame (number of rows)

Python's built-in len( ) function returns the number of rows when a data frame is passed as an argument. 

In [2]:
len(store_df)

541909

## Using dots ( . ) instead of brackets ( [ ] ) to select columns and rows

The main rule is that the column name cannot contain any spaces. If it does, you cannot select it with ( . ). You have to use [ ] and type the column name as a string

In [3]:
# Using brackets
store_df["Country"]

0         United Kingdom
1         United Kingdom
2         United Kingdom
3         United Kingdom
4         United Kingdom
               ...      
541904            France
541905            France
541906            France
541907            France
541908            France
Name: Country, Length: 541909, dtype: object

In [4]:
# Using a dot
store_df.Country

0         United Kingdom
1         United Kingdom
2         United Kingdom
3         United Kingdom
4         United Kingdom
               ...      
541904            France
541905            France
541906            France
541907            France
541908            France
Name: Country, Length: 541909, dtype: object

## View data types

It is almost always necessary to take a look at the data types of columns before moving on with your analysis. Here, object data types stand for strings. 

In [5]:
store_df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

## Select columns based on data types

You can select columns based on data types. In the following example, we select only the integer columns and see that there is only one integer column. 

In [6]:
# Viewing only integer columns
store_df.select_dtypes(include="int64")

Quantity
0              6
1              6
2              8
3              6
4              6
...          ...
541904        12
541905         6
541906         4
541907         4
541908         3

[541909 rows x 1 columns]

In [7]:
# Viewing only float columns
store_df.select_dtypes(include=float) 

UnitPrice  CustomerID
0            2.55     17850.0
1            3.39     17850.0
2            2.75     17850.0
3            3.39     17850.0
4            3.39     17850.0
...           ...         ...
541904       0.85     12680.0
541905       2.10     12680.0
541906       4.15     12680.0
541907       4.15     12680.0
541908       4.95     12680.0

[541909 rows x 2 columns]

## Check for NaN (missing) values

NaN values stand for missing values. They are unidentified values. 

We will have a whole part dedicated to dealing with NaN (missing) values. It is a delicate subject. For now, let's see how we can use Pandas to take a look at how many of them we have in our data set. 

In [8]:
# Returns true if the column contains any missing values. 
store_df.isna().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
dtype: bool

In [9]:
# Total number of missing values in each column
store_df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

## Value counts 

We can see how many different types of values were observed in a column using the following method:


In [10]:
# Number of times each country was observed in the 'Country' column
store_df.Country.value_counts()

United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

## Largest & smallest values

We can use nlargest and nsmallest methods to get the largest and smallest values in a data frame by column. This is a very practical method, as we may often want to have a look at top or bottom values of a column. 

In [11]:
# Entries with 10 largest unit prices
store_df.nlargest(10, "UnitPrice")

InvoiceNo  StockCode      Description  Quantity    InvoiceDate  \
222681   C556445          M           Manual        -1  6/10/11 15:31   
524602   C580605  AMAZONFEE       AMAZON FEE        -1  12/5/11 11:36   
43702    C540117  AMAZONFEE       AMAZON FEE        -1    1/5/11 9:55   
43703    C540118  AMAZONFEE       AMAZON FEE        -1    1/5/11 9:57   
15016    C537630  AMAZONFEE       AMAZON FEE        -1  12/7/10 15:04   
15017     537632  AMAZONFEE       AMAZON FEE         1  12/7/10 15:08   
16356    C537651  AMAZONFEE       AMAZON FEE        -1  12/7/10 15:49   
16232    C537644  AMAZONFEE       AMAZON FEE        -1  12/7/10 15:34   
524601   C580604  AMAZONFEE       AMAZON FEE        -1  12/5/11 11:35   
299982   A563185          B  Adjust bad debt         1  8/12/11 14:50   

        UnitPrice  CustomerID         Country  
222681   38970.00     15098.0  United Kingdom  
524602   17836.46         NaN  United Kingdom  
43702    16888.02         NaN  United Kingdom  
43703    16453.71         NaN  United Kingdom  
15016    13541.33         NaN  United Kingdom  
15017    13541.33         NaN  United Kingdom  
16356    13541.33         NaN  United Kingdom  
16232    13474.79         NaN  United Kingdom  
524601   11586.50         NaN  United Kingdom  
299982   11062.06         NaN  United Kingdom

In [12]:
# Entries with 10 smallest unit prices
store_df.nsmallest(10, "UnitPrice")

InvoiceNo StockCode      Description  Quantity    InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/11 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/11 14:52   
622       536414     22139              NaN        56  12/1/10 11:52   
1970      536545     21134              NaN         1  12/1/10 14:32   
1971      536546     22145              NaN         1  12/1/10 14:33   
1972      536547     37509              NaN         1  12/1/10 14:33   
1987      536549    85226A              NaN         1  12/1/10 14:34   
1988      536550     85044              NaN         1  12/1/10 14:34   
2024      536552     20950              NaN         1  12/1/10 14:34   
2025      536553     37461              NaN         3  12/1/10 14:35   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom  
622          0.00         NaN  United Kingdom  
1970         0.00         NaN  United Kingdom  
1971         0.00         NaN  United Kingdom  
1972         0.00         NaN  United Kingdom  
1987         0.00         NaN  United Kingdom  
1988         0.00         NaN  United Kingdom  
2024         0.00         NaN  United Kingdom  
2025         0.00         NaN  United Kingdom

## Query the data frame

Query allows us to filter a data frame based on certain condition(s).

In [13]:
# Observations where Quantity is larger than Unit Price
store_df.query("Quantity > UnitPrice")

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541900    581587     22730           ALARM CLOCK BAKELIKE IVORY         4   
541901    581587     22367      CHILDRENS APRON SPACEBOY DESIGN         8   
541902    581587     22629                  SPACEBOY LUNCH BOX         12   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
541900  12/9/11 12:50       3.75     12680.0          France  
541901  12/9/11 12:50       1.95     12680.0          France  
541902  12/9/11 12:50       1.95     12680.0          France  
541904  12/9/11 12:50       0.85     12680.0          France  
541905  12/9/11 12:50       2.10     12680.0          France  

[315797 rows x 8 columns]

## Check for unique values and their number

In [14]:
store_df.Country.unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [15]:
# Number of unique value
store_df.Country.nunique()

38

## Use a list to check for specific values

We can define a list with the values that are important for us. Then, we can use the .isin( ) method to quickly filter a column for the values in the list.

In [16]:
# Define a list
list_of_countries = ['United Kingdom', 'Germany', 'USA', 'Canada']

# Check the column for the values in the list
store_df[store_df['Country'].isin(list_of_countries)]

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541889    581585     22466       FAIRY TALE COTTAGE NIGHT LIGHT        12   
541890    581586     22061  LARGE CAKE STAND  HANGING STRAWBERY         8   
541891    581586     23275     SET OF 3 HANGING OWLS OLLIE BEAK        24   
541892    581586     21217        RED RETROSPOT ROUND CAKE TINS        24   
541893    581586     20685                DOORMAT RED RETROSPOT        10   

          InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/10 8:26       2.55     17850.0  United Kingdom  
1        12/1/10 8:26       3.39     17850.0  United Kingdom  
2        12/1/10 8:26       2.75     17850.0  United Kingdom  
3        12/1/10 8:26       3.39     17850.0  United Kingdom  
4        12/1/10 8:26       3.39     17850.0  United Kingdom  
...               ...        ...         ...             ...  
541889  12/9/11 12:31       1.95     15804.0  United Kingdom  
541890  12/9/11 12:49       2.95     13113.0  United Kingdom  
541891  12/9/11 12:49       1.25     13113.0  United Kingdom  
541892  12/9/11 12:49       8.95     13113.0  United Kingdom  
541893  12/9/11 12:49       7.08     13113.0  United Kingdom  

[505415 rows x 8 columns]

## Replace values

We can use the replace method to change the values of a column. The inplace= parameter allows us to make the changes permanent when we set it to True.

In [17]:
# Replace "United Kingdom" with "UK", make the change permanent by setting inplace to True
store_df.Country.replace({"United Kingdom":"UK"}, inplace=True)
store_df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID Country  
0        12/1/10 8:26       2.55     17850.0      UK  
1        12/1/10 8:26       3.39     17850.0      UK  
2        12/1/10 8:26       2.75     17850.0      UK  
3        12/1/10 8:26       3.39     17850.0      UK  
4        12/1/10 8:26       3.39     17850.0      UK  
...               ...        ...         ...     ...  
541904  12/9/11 12:50       0.85     12680.0  France  
541905  12/9/11 12:50       2.10     12680.0  France  
541906  12/9/11 12:50       4.15     12680.0  France  
541907  12/9/11 12:50       4.15     12680.0  France  
541908  12/9/11 12:50       4.95     12680.0  France  

[541909 rows x 8 columns]

## Rename Columns

We can rename columns in a way similar to how we use the .replace method. 

In [18]:
store_df.rename(columns={'Description':'ItemDescr'}, inplace=True)
store_df

InvoiceNo StockCode                            ItemDescr  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID Country  
0        12/1/10 8:26       2.55     17850.0      UK  
1        12/1/10 8:26       3.39     17850.0      UK  
2        12/1/10 8:26       2.75     17850.0      UK  
3        12/1/10 8:26       3.39     17850.0      UK  
4        12/1/10 8:26       3.39     17850.0      UK  
...               ...        ...         ...     ...  
541904  12/9/11 12:50       0.85     12680.0  France  
541905  12/9/11 12:50       2.10     12680.0  France  
541906  12/9/11 12:50       4.15     12680.0  France  
541907  12/9/11 12:50       4.15     12680.0  France  
541908  12/9/11 12:50       4.95     12680.0  France  

[541909 rows x 8 columns]

## Check for changes from previous values

The .pct_change( ) method returns the amount by which a row has changed when compared with the previous row, in percentage. 

The first row will always be NaN because there are no rows that come before it for comparison.

In [19]:
# Change in Unit Price per row
store_df.UnitPrice.pct_change()

0              NaN
1         0.329412
2        -0.188791
3         0.232727
4         0.000000
            ...   
541904   -0.795181
541905    1.470588
541906    0.976190
541907    0.000000
541908    0.192771
Name: UnitPrice, Length: 541909, dtype: float64

## Use pivot tables

For this example we will use the automobile dataset which we have also used in the previous chapter. Let's not forget that the data frame contains some entries marked with '?' and we have to turn them into NaN before we move on

In [20]:
# Load the data frame, turn '?' entries into NaN
auto_df = pd.read_csv('auto-mpg.csv', na_values='?')

# Replace region numbers with actual region names
auto_df.origin.replace({1:'USA', 2:'Europe', 3:'Asia'}, inplace=True)
auto_df


mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model year  origin                   car name  
0            70     USA  chevrolet chevelle malibu  
1            70     USA          buick skylark 320  
2            70     USA         plymouth satellite  
3            70     USA              amc rebel sst  
4            70     USA                ford torino  
..          ...     ...                        ...  
393          82     USA            ford mustang gl  
394          82  Europe                  vw pickup  
395          82     USA              dodge rampage  
396          82     USA                ford ranger  
397          82     USA                 chevy s-10  

[398 rows x 9 columns]

In [21]:
# Average (mean) horsepower for cylinders, grouped by regions of origin
pd.pivot_table(data=auto_df, index='origin', values = 'horsepower', columns='cylinders', aggfunc=np.mean)

cylinders      3          4          5           6           8
origin                                                        
Asia       99.25  75.579710        NaN  115.833333         NaN
Europe       NaN  78.311475  82.333333  113.500000         NaN
USA          NaN  80.956522        NaN   99.671233  158.300971

Let's read and understand this table together:

For Asia:
- Cars with 3 cylinders have an average of 99.25 horsepower
- Cars with 4 cylinders have an average of 75.57 horsepower
- There are no cars with 5 cylinders
- Cars with 6 cylinders have an average of 115.83 horsepower
- There are no cars with 8 cylinders

You can figure out the rest on your own.

Please note that such tables are meant to give you an idea, not the whole picture. There may be underlying reasons for the abnormalities you observe in the tables. For example, there could be reason for 3 cylinders Asian cars having an higher average of horsepower when compared to cars with 4 cylinders. Maybe, there were more observations for cars with 3 cylinders and this resulted in a higher average of horsepower. It is always good to be careful when carrying out data analysis even when the tools we have seem to make it so easy.

### Exercise


* Use the functions and methods shown here on 3 different data sets of your choice.
* Prepare 2 pivot tables on different data sets. Use a different aggregation function (aggfunc) for each.